### PR TITLE
improved parsing error messages

### LINKED ISSUE #204: Improve datatype parsing error message
Currently errors that occur during datatype parsing are often confusing or just not very good. This needs to be improved.

Addions by @bigerl:
The `ConstexprString datatype` parameter in some parsing functions does not enforce in any way to be set to correct values only. It's very easy to set it to some random value. I have no simple idea at hand but a more robust approach would be preferable.

In general a more robust way to generate messages would be preferable.

### PR DESCRIPTION
**Pull request by mcb5637**
close #204

### CODE REVIEW

File: src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp [hunk 1]
```diff
@@ -65,26 +37,53 @@ bool to_chars_canonical(I const value, writer::BufWriterParts const writer) noex
...
+    if (res.ec != std::errc{}) {
+        if (res.ec == std::errc::invalid_argument) {
+            bool neg = true;
+            if constexpr(!std::unsigned_integral<F>) {
+                neg = *res.ptr != '-';
+            }
+            if(res.ptr != s.data() + s.size() && neg) {
+                // fallthrough to data check
+            }
+            else {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp]: formatting

File: src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp [hunk 2]
```diff
@@ -65,26 +37,53 @@ bool to_chars_canonical(I const value, writer::BufWriterParts const writer) noex
...
+                throw rdf4cpp::InvalidNode{std::format("{} parsing error: literal is empty", datatype)};
+            }
+        }
+        else if (res.ec == std::errc::result_out_of_range) {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp]: formatting

File: src/rdf4cpp/datatypes/xsd/time/Time.cpp [hunk 3]
```diff
@@ -10,19 +10,19 @@ capabilities::Default<xsd_time>::cpp_type capabilities::Default<xsd_time>::from_
     using namespace registry::util;
     auto hours = parse_date_time_fragment<std::chrono::hours, unsigned int, ':', identifier>(s);
     auto minutes = parse_date_time_fragment<std::chrono::minutes, unsigned int, ':', identifier>(s);
-    auto tz = rdf4cpp::Timezone::parse_optional(s);
+    auto tz = rdf4cpp::Timezone::parse_optional(s, identifier);
     std::chrono::milliseconds ms = parse_milliseconds<identifier>(s);
     if (!registry::relaxed_parsing_mode) {
         if (minutes < std::chrono::minutes(0) || minutes > std::chrono::hours(1))
-            throw InvalidNode{"minutes out of range"};
+            throw InvalidNode(std::format("{} parsing error: minutes out of range", identifier));
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/Time.cpp]: curly braces

File: src/rdf4cpp/datatypes/xsd/time/DayTimeDuration.cpp [hunk 4]
```diff
@@ -15,15 +15,15 @@ capabilities::Default<xsd_dayTimeDuration>::cpp_type capabilities::Default<xsd_d
...
+        throw InvalidNode{std::format("{} parsing error: duration missing P", identifier)};
     s = s.substr(1);
     auto p = s.find('T');
     auto date = s.substr(0, p);
     auto time = p == std::string::npos ? std::string_view{""} : s.substr(p);
     auto days = parse_duration_fragment<std::chrono::days, uint64_t, 'D', identifier>(date);
     if (!time.empty()) {
         if (time[0] != 'T')
-            throw InvalidNode{"duration missing T"};
+            throw InvalidNode{std::format("{} parsing error: duration missing T", identifier)};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DayTimeDuration.cpp:27]: curly braces

File: src/rdf4cpp/datatypes/xsd/time/DayTimeDuration.cpp [hunk 5]
```diff
@@ -15,15 +15,15 @@ capabilities::Default<xsd_dayTimeDuration>::cpp_type capabilities::Default<xsd_d
         s = s.substr(1);
     }
     if (s.empty() || s[0] != 'P')
-        throw InvalidNode{"duration missing P"};
+        throw InvalidNode{std::format("{} parsing error: duration missing P", identifier)};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DayTimeDuration.cpp:18]: curly braces

File: src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp [hunk 6]
```diff
@@ -15,34 +15,34 @@ capabilities::Default<xsd_dateTimeStamp>::cpp_type capabilities::Default<xsd_dat
...
-            throw InvalidNode{"seconds out of range"};
+            throw InvalidNode(std::format("{} parsing error: seconds out of range", identifier));
     }
     auto time = hours + minutes + ms;
     if (!registry::relaxed_parsing_mode) {
         if (time == std::chrono::hours{24}) {
             date = std::chrono::year_month_day{std::chrono::local_days{date} + std::chrono::days{1}};
             if (!date.ok())
-                throw InvalidNode("invalid date");
+                throw InvalidNode(std::format("{} parsing error: {:%Y-%m-%d} is invalid", identifier, date));
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp:48]: curly braces

File: src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp [hunk 7]
```diff
@@ -15,34 +15,34 @@ capabilities::Default<xsd_dateTimeStamp>::cpp_type capabilities::Default<xsd_dat
...
     if (!registry::relaxed_parsing_mode) {
         if (minutes < std::chrono::minutes(0) || minutes > std::chrono::hours(1))
-            throw InvalidNode{"minutes out of range"};
+            throw InvalidNode(std::format("{} parsing error: minutes out of range", identifier));
         if (hours < std::chrono::hours(0) || hours > std::chrono::days(1))
-            throw InvalidNode{"hours out of range"};
+            throw InvalidNode(std::format("{} parsing error: hours out of range", identifier));
         if (ms < std::chrono::seconds(0) || ms > std::chrono::minutes(1))
-            throw InvalidNode{"seconds out of range"};
+            throw InvalidNode(std::format("{} parsing error: seconds out of range", identifier));
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/time/DateTimeStamp.cpp:40]: curly braces

File: src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp [hunk 8]
```diff
@@ -65,26 +37,53 @@ bool to_chars_canonical(I const value, writer::BufWriterParts const writer) noex
...
+        }
+    }();
 
-    if (res.ec != std::errc{} || res.ptr != s.data() + s.size()) {
+    if (res.ec != std::errc{}) {
+        if (res.ec == std::errc::invalid_argument) {
+            bool neg = true;
+            if constexpr(!std::unsigned_integral<F>) {
+                neg = *res.ptr != '-';
+            }
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp:69]: add newline after this

File: src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp [hunk 9]
```diff
@@ -65,26 +37,53 @@ bool to_chars_canonical(I const value, writer::BufWriterParts const writer) noex
...
+requires (std::floating_point<F> || std::integral<F>)
 F from_chars(std::string_view s) {
     if (s.starts_with('+')) {
         // from_chars does not allow initial +
         s.remove_prefix(1);
     }
 
     F value;
-    std::from_chars_result const res = std::from_chars(s.data(), s.data() + s.size(), value, std::chars_format::general);
+    std::from_chars_result res = [&]() {
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/registry/util/CharConvExt.hpp]: noexcept

**Review by liss-h — CHANGES_REQUESTED** [review]
Thanks
Apart from some formatting things that we might as well fix now, I have nothing to complain (I've not marked all of them, you can just look at the diff to find more of them).

The changes should make the error messages quite a bit better

**Comment by liss-h** [comment]
Converting to draft until you've looked into the decimal stuff

**Comment by mcb5637** [comment]
concidering the time that would be needed for more precise error messages, the existing boost messages with added datatype are good enough. result is something like this: `xsd:integer parsing error: Unexpected character encountered in input.` (for any non numeric char in the input)

**Review by liss-h — CHANGES_REQUESTED** [review]
A more general comment: I think it would look more "professional" if the error messages were consistent with respect to how datatypes are called. Currently there seem to be both short form (`xsd:xyz`) and full iri `http://...#xyz`. I think it would be best to choose and stick to one of them (iri probably makes things easier)

File: src/rdf4cpp/datatypes/xsd/integers/non_negative/NonNegativeInteger.cpp [hunk 10]
```diff
@@ -14,11 +14,11 @@ capabilities::Default<xsd_non_negative_integer>::cpp_type capabilities::Default<
     try {
         ret = cpp_type{s};
     } catch (std::runtime_error const &e) {
-        throw InvalidNode{std::string{"xsd:nonNegativeInteger parsing error: "} + e.what()};
+        throw InvalidNode{std::string{"http://www.w3.org/2001/XMLSchema#nonNegativeInteger parsing error: "} + e.what()};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/integers/non_negative/NonNegativeInteger.cpp]: why not use the `identifier`?

File: src/rdf4cpp/datatypes/xsd/integers/non_negative/PositiveInteger.cpp [hunk 11]
```diff
@@ -12,11 +12,11 @@ capabilities::Default<xsd_positive_integer>::cpp_type capabilities::Default<xsd_
     try {
         ret = cpp_type{s};
     } catch (std::runtime_error const &e) {
-        throw InvalidNode{std::string{"xsd:positiveInteger parsing error: "} + e.what()};
+        throw InvalidNode{std::string{"http://www.w3.org/2001/XMLSchema#positiveInteger parsing error: "} + e.what()};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/integers/non_negative/PositiveInteger.cpp]: same here

File: src/rdf4cpp/datatypes/owl/Rational.cpp [hunk 12]
```diff
@@ -20,18 +20,18 @@ template<>
 capabilities::Default<owl_rational>::cpp_type capabilities::Default<owl_rational>::from_string(std::string_view s) {
     if (auto pos = s.find_last_of('-'); pos != std::string_view::npos && pos != 0) {
         // owl:rational only allows - at beginning, boost also allows it in the denominator
-        throw InvalidNode{"owl:rational parsing error: invalid sign position"};
+        throw InvalidNode{"http://www.w3.org/2002/07/owl#rational parsing error: invalid sign position"};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/owl/Rational.cpp]: same here

File: src/rdf4cpp/datatypes/owl/Real.cpp [hunk 13]
```diff
@@ -9,7 +9,7 @@ capabilities::Default<owl_real>::cpp_type capabilities::Default<owl_real>::from_
     try {
         return cpp_type{s};
     } catch (std::runtime_error const &e) {
-        throw InvalidNode{std::string{"owl:real parsing error: "} + e.what()};
+        throw InvalidNode{std::string{"http://www.w3.org/2002/07/owl#real parsing error: "} + e.what()};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/owl/Real.cpp]: same here

File: src/rdf4cpp/datatypes/xsd/integers/signed/Integer.cpp [hunk 14]
```diff
@@ -16,7 +16,7 @@ capabilities::Default<xsd_integer>::cpp_type capabilities::Default<xsd_integer>:
     try {
         return cpp_type{s};
     } catch (std::runtime_error const &e) {
-        throw InvalidNode{std::string{"xsd:integer parsing error: "} + e.what()};
+        throw InvalidNode{std::string{"http://www.w3.org/2001/XMLSchema#integer parsing error: "} + e.what()};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/integers/signed/Integer.cpp]: same here

File: src/rdf4cpp/datatypes/xsd/integers/non_positive/NonPositiveInteger.cpp [hunk 15]
```diff
@@ -14,11 +14,11 @@ capabilities::Default<xsd_non_positive_integer>::cpp_type capabilities::Default<
     try {
         ret = cpp_type{s};
     } catch (std::runtime_error const &e) {
-        throw InvalidNode{std::string{"xsd:nonPositiveInteger parsing error: "} + e.what()};
+        throw InvalidNode{std::string{"http://www.w3.org/2001/XMLSchema#nonPositiveInteger parsing error: "} + e.what()};
```
Comment (liss-h) [inline src/rdf4cpp/datatypes/xsd/integers/non_positive/NonPositiveInteger.cpp]: same here

**Review by liss-h — CHANGES_REQUESTED** [review]
There are a few places where you could have used the `identifier` as far as I can tell, but didn't. Idk if I missed something but if not please fix those

**Review by liss-h — APPROVED** [review]

**Pull request merged**